commit 5b6c49ed5c8d
Author: James Teh <jteh@mozilla.com>
Date:   Thu Apr 12 16:32:19 2018 +1000

    Bug 1453555: Fix accessibility group info for <select size="1"> options. r=surkov a=jcristau
    
    In the e10s implementation, Accessible::NativeState for the options doesn't include the invisible state. (It does with e10s disabled.)
    In HTMLSelectOptionAccessible::NativeState, rather than just flipping (xor) the invisible state, absolutely ensure it gets removed. We don't want to *add* the invisible state if it isn't there.
    This allows group position info to be calculated correctly.
    
    MozReview-Commit-ID: LPEVhOOm2NT
    
    --HG--
    extra : source : f52a56dc215cfd8ea7310c168632d33531fc8c90
---
 accessible/html/HTMLSelectAccessible.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git accessible/html/HTMLSelectAccessible.cpp accessible/html/HTMLSelectAccessible.cpp
index 0cbe7ac89dce..efa05c1adfc1 100644
--- accessible/html/HTMLSelectAccessible.cpp
+++ accessible/html/HTMLSelectAccessible.cpp
@@ -198,7 +198,10 @@ HTMLSelectOptionAccessible::NativeState()
     // visible option
     if (!selected) {
       state |= states::OFFSCREEN;
-      state ^= states::INVISIBLE;
+      // Ensure the invisible state is removed. Otherwise, group info will skip
+      // this option. Furthermore, this gets cached and this doesn't get
+      // invalidated even once the select is expanded.
+      state &= ~states::INVISIBLE;
     } else {
       // Clear offscreen and invisible for currently showing option
       state &= ~(states::OFFSCREEN | states::INVISIBLE);
